File: 2552-maximum-sum-of-distinct-subarrays-with-length-k/maximum-sum-of-distinct-subarrays-with-length-k.java
class Solution {
    public long maximumSubarraySum(int[] nums, int k) {
        Map<Integer, Integer> map = new HashMap<>();
        long sum = 0, maxSum = 0;
        int i = 0;
        for(int j = 0; j < nums.length; j++) {
            sum += nums[j];
            map.put(nums[j], map.getOrDefault(nums[j], 0) + 1);
            if(j - i + 1 == k) {
                if(map.size() == k) {
                    maxSum = Math.max(maxSum, sum);
                }
                sum -= nums[i];
                map.put(nums[i], map.getOrDefault(nums[i], 0) - 1);
                if(map.get(nums[i]) == 0) {
                    map.remove(nums[i]);
                }
                i++;
            }
        }
        return maxSum;
    }
}

File: 2552-maximum-sum-of-distinct-subarrays-with-length-k/maximum-sum-of-distinct-subarrays-with-length-k.cpp
class Solution {
public:
    long long maximumSubarraySum(vector<int>& nums, int k) 
    {
        unordered_map<int, int> map;
        int i = 0;
        long long sum = 0, maxSum = 0;
        for (int j = 0; j < nums.size(); ++j) 
        {
            sum += nums[j];
            map[nums[j]]++;
            if(j - i + 1 == k) 
            {
                if(j - i + 1 == map.size())
                maxSum = max(maxSum, sum);

                sum -= nums[i];
                map[nums[i]]--;

                if(!map[nums[i]])
                map.erase(nums[i]);
                i++;
            }
        }
        return maxSum;
    }
};